sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs

```diff
@@ -184,88 +100,185 @@ internal virtual void Wait(TimeSpan time, CancellationToken cancellationToken)
+
+        internal virtual async Task WaitAsync(TimeSpan time, CancellationToken cancellationToken)
+        {
+            await Task.Delay(time, cancellationToken).ConfigureAwait(false);
+        }
+
+        internal virtual void Wait(TimeSpan time, CancellationToken cancellationToken)
+        {
+            cancellationToken.WaitHandle.WaitOne(time);
+        }
 
-        private TimeSpan GetNextDelayInternal(HttpMessage message)
+        /// <summary>
+        /// This type implements a System.ClientModel
+        /// <see cref="ClientRetryPolicy"/> and adds the Azure.Core-specific
+        /// feature of creating an EventSource event with the elapsed time to
+        /// process the message on each try.
+        ///
+        /// It also adapts the Azure.Core <see cref="RetryPolicy"/> that holds
+        /// it as a member. This is needed so that, if a user has implemented a
+        /// type derived from Azure.Core <see cref="RetryPolicy"/> and
+        /// overridden one or more of its virtual methods, when a virtual method
+        /// is called on the <see cref="ClientRetryPolicy"/> base type from a
+        /// System.ClientModel context, it will call through to the overridden
+        /// method on the derived type.
+        /// </summary>
+        private sealed class RetryPolicyAdapter : ClientRetryPolicy
         {
-            return _delayStrategy.GetNextDelay(
-                message.HasResponse ? message.Response : default,
-                message.RetryNumber + 1);
+            private readonly RetryPolicy _azureCorePolicy;
+            private readonly DelayStrategy _delayStrategy;
+
+            public RetryPolicyAdapter(int maxRetries, DelayStrategy delay, RetryPolicy policy)
+                : base(maxRetries)
+            {
+                _delayStrategy = delay;
+                _azureCorePolicy = policy;
+            }
+
+            protected override void OnSendingRequest(PipelineMessage message)
+            {
+                message.SetProperty(typeof(BeforeTimestamp), Stopwatch.GetTimestamp());
+
+                _azureCorePolicy.OnSendingRequest(HttpMessage.GetHttpMessage(message));
+            }
+
+            protected override async ValueTask OnSendingRequestAsync(PipelineMessage message)
+            {
+                message.SetProperty(typeof(BeforeTimestamp), Stopwatch.GetTimestamp());
+
+                await _azureCorePolicy.OnSendingRequestAsync(HttpMessage.GetHttpMessage(message)).ConfigureAwait(false);
+            }
+
+            protected override void OnRequestSent(PipelineMessage message)
+            {
+                _azureCorePolicy.OnRequestSent(HttpMessage.GetHttpMessage(message));
+
+                if (!message.TryGetProperty(typeof(BeforeTimestamp), out object? beforeTimestamp) ||
+                    beforeTimestamp is not long before)
+                {
+                    throw new InvalidOperationException("'BeforeProcessTimestamp' was not set on message by RetryPolicy.");
```

Comment:
Is this possible? Would it be better to log here than to throw - I think I'd rather have the retry execute without timing data than to hard fail.

Reply:
That's a good point.  I don't think it is, but I'll change it to a Debug.Assert and reset the values to something that makes more sense.  Thanks!